README.md
# 6oz.js

A JavaScript library using intuitive templating and Google's Incremental DOM library to efficiently update data changes to the real browser DOM.

## Getting Started

You will need to download and place both the Incremental DOM and 6oz.js JavaScript files on your website.  Load Incremental DOM first.  The copy of Incremental DOM in this repository has a bug fix that sets the elements actual properties for attributes like `value`, `checked`, `disabled`, etc.

```
<script type="text/javascript" src="incremental-dom-min.js"></script>
<script type="text/javascript" src="6oz.min.js"></script>
```

## Templating

Templating is done using code blocks.  Within the code blocks you can use normal JavaScript.  Simple!  The code blocks are inspired by and match to ASP.Net Web Forms code blocks.  To start a code block write `<%` and to close a code block write `%>`.


#### Rendering Data
To render data in your template that is not HTML encoded use the start block `<%=`.  Use this if you wish to write in HTML into the template.

To render data in your template that is HTML encoded use the start block `<%:`.  Use this if you wish to write in text into the template.

For the examples below we are using the following variable
```
var templateData = {
	"htmlText": "<p>Hello world</p>"
};
```
**Template 1**
```
<div><%=htmlText %></div>
```
*will yield*
```
<div><p>Hello world</p></div>
```
**Template 2**
```
<div><%:htmlText %></div>
```
*will yield*
```
<div>&lt;p&gt;Hello world&lt;p&gt;</div>
```

#### JavaScript In Your Template
You can run JavaScript right in your template.  This is useful to conditionally render and/or loop through data in your render logic.

For the example below we are using the following variable
```
var templateData = {
	"header": "My Example",
	"showHeader": true,
	"items": [
		"Hi there",
		"Hello mate",
		"Gidday",
		"Mate",
		"Dave"
	]
};
```
**Template**
```
<% if (showHeader) { %>
<h1><%:header %></h1>
<% } %>
<ul>
<% for (var i = 0, l = items.length; i < l; i++) { %>
	<li><%:items[i] %></li>
<% } %>
</ul>
```
*will yield*
```
<h1>My Example</h1>
<ul>
	<li>Hi there</li>
	<li>Hello mate</li>
	<li>Gidday</li>
	<li>Mate</li>
	<li>Dave</li>
</ul>
```
## Incremental DOM
6oz.js uses Google's Incremental DOM JavaScript library to efficiently patch rendered templates against the DOM.  The low level complexities are all handled by 6oz.js.

To start call the `applyToDOM(mountElement, templateOrTemplateID, data)` function with three arguments.  This will return a context to allow you to easily update the template with new/changed data.

```
var elMount = document.getElementById("app");
var template = '<div onclick="demoTemplate_click" class="demo-template"><%:text %></div>';
var demoTemplate_click = function () {
	alert("Demo template clicked!");
};
var templateData = {
	"data": {
		"text": "Hello world"
	},
	"functions": {
		"demoTemplate_click": demoTemplate_click
	}
};
var context = __6oz.applyToDOM(elMount, template, templateData);

templateData.data.text = "New text will be updated";
context.update(templateData);
```

##### Lists (key-id)
Within your templates you may have items that are repeated such as list items (li) or select options (option).  Incremental DOM can have a hard time discerning the differences when updating if the items are similar.  To aid the diffing process, add an attribute of `key-id` on your repeating element.  More about this here, http://google.github.io/incremental-dom/#conditional-rendering/array-of-items.

##### Props
Introduced in v1.0.1 you can now attach data properties to an element by using the attribute `props` on your element.  This is a special attribute for 6oz and it will attach the data to that element.  This should be used in place of storing literal values on the DOM itself.  You can store any JavaScript object using this.  i.e. string, bool, numbers, functions, object.

You can access the properties attached in any event function(s) on that element.  It is accessible via `this.props`.

_Below is template code_
```
<div id={elementID} props={myDivProperties} onclick={div_click}></div>
```
_Below is JavaScript code_
```
var data = {
	"elementID": "divHelloWorld",
	"myDivProperties": {
		"isTest": true,
		"myCustomFunction": function () {
			console.log("Hello world!");
		},
		"myInt": 1000
	},
	"div_click": function () {
		console.log("This is how you access your properties in events", this.props);
	}
};
```

##### Ignoring elements (skip-node)
When using external libraries that mutate the DOM, 6oz will need to ignore the parent element that the external library is contained in.  If this isn't applied, when 6oz updates the DOM it will override the changes to the DOM.  To ignore a DOM element, add `skip-node="true"` to the element.  It will ignore the element you apply the attribute to and all it's children.

```
<div>
	<h1>This is the header for the template</h1>
	<p><%:paragraphText %></p>
	<div class="external-lookup" skip-node="true"></div>
</div>
```

## Components
6oz.js allows you to create components which are templates that can be reused.  You can define components using the snippet below.  The component name must be kebab case and contain at least one hyphen.  The second parameter must be the template or the template ID.

```
__6oz.addComponent({
    "componentName": "my-custom-component-name",
    "template": "<h1>This is my template</div>",
    "controller": function (props) {
        this.myFunctionInsideComponent = function () {
            console.log("myFunctionInsideComponent controller function invoked");
        };
    } 
});
__6oz.addComponent({
    "componentName": "my-custom-component-name-2",
    "template": "myTemplateID"
});
```

Use the components in any of your templates (including component templates, components in components baby!) in your app by writing a tag with your component name.  You can pass data to be used in the components using attributes on the tag.  If it's not text, use curly braces to pass JavaScript objects/data through.

_Below is template code_
```
<% var templateDataInScope = {
    "arrayOfStuff": [1, 2, 3],
    "o": {
        "value": "Hello world"
    }
}; %>
<my-custom-component-name
    bodyText="This is an example of just passing a string in."
    customData={templateDataInScope}
    hardCodedData={{ "o": "Fixed data" }} />
```

Accessing the data passed into the component within the template is simple.  Just use the `props` object in the template to access the data.  You can also access the controller data and functions by using the `controller` object in the template.

```
<h1 onclick="controller.myControllerFunction">This is the my component template</h1>
<p><%=props.bodyText %></p>
<pre><%=props.customData.o.value %></pre>
<ul>
<% var arr = props.customData.arrayOfStuff;
for (var i = 0, l = arr.length; i < l; i++) { %>
<li><%=arr[i] %></li>
<% } %>
</ul>
```

## FAQ
### Unusual rendering and event bindings are occurring
This can occur when the Incremental DOM library cannot discern the difference between two elements in updates.  This can occur if the DOM structure is similar in two different templates.  To resolve this, place a key-id on the top level element that is having issues.  Refer to Lists (key-id) above in README.

## Authors

**Phillip Moon**

## License

This project is licensed under the MIT License - see the [LICENSE.md](LICENSE.md) file for details


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Released]

## [1.0.2] - 2018-12-10
### Added
- Added ability to skip nodes.  To be used for external libraries.

### Changed
- Changed the lex and templating to escape values correctly.  Fixes bug.

### Removed
- Nothing.

## [1.0.1] - 2018-04-22
### Added
- Added changelog file.
- Added ability to attach properties to an element.  This is achievable by using the `prop` attribute.

### Changed
- Changed data bridge logic to check that curly braces are used on an element attribute.  This fixes an issue where the using curly braces else where in the template caused an exception.  It's an improvement, but still not fool proof.  Will have to work up a better regex at some point.

### Removed
- Nothing.

test/app.js
/**
 * BASIC APP
 */
(function () {
	var templateData = {
		"data": {
			"paragraph": "This is a test paragraph",
			"heading": "Hello World",
			"now": new Date().toString(),
			"inputValue": "",
			"d": {
				"x": 0,
				"y": 0
			}
		},
		"functions": {
			"fireme": fireme,
			"datetime_update": datetime_update,
			"input_update": input_update,
			"d_mousedown": d_mousedown,
			"d_mousemove": d_mousemove,
			"d_mouseup": d_mouseup
		}
	};
	var template = '<div onmousedown="d_mousedown" onmousemove="d_mousemove" onmouseup="d_mouseup" style="position:absolute;transform:translate3d(<%=d.x %>,<%=d.y %>, 0);width:20px;height:20px;border-radius:50%;background-color:red;"></div><div class="true-stuff-class no-show"><h1><%:heading %></h1><input style="width: 400px;" type="text" onkeyup="input_update" value="<%:inputValue %>" placeholder="Hideho" /><p onclick="fireme"><%:paragraph %></p><div class="row" onmouseover="datetime_update">Row<div class="col">Col</div></div><pre>Time: <%:now %></pre></div>';

	__6oz.applyToDOM(document.getElementById("basicApp"), template, templateData);

	function fireme(e) {
		alert("You're fired!");
	}

	function datetime_update() {
		templateData.data.now = new Date().toString();
		__6oz.applyToDOM(document.getElementById("basicApp"), template, templateData);
	}

	function input_update(e) {
		templateData.data.inputValue = this.value;
		templateData.data.heading = this.value;
		__6oz.applyToDOM(document.getElementById("basicApp"), template, templateData);
	}

	var startDragPos = {};
	var isMouseDown = false;

	function d_mousedown(e) {
		startDragPos = { "x": e.clientX, "y": e.clientY };
		isMouseDown = true;
		console.log("d_mousedown", e);
	}

	function d_mousemove(e) {
		console.log("d_mousemove", e);

		if (isMouseDown) {
			templateData.data.d.x = (e.clientX - startDragPos.x) + "px";
			templateData.data.d.y = (e.clientY - startDragPos.y) + "px";
			console.log(templateData.data.d);
			__6oz.applyToDOM(document.getElementById("basicApp"), template, templateData);
		}
	}

	function d_mouseup(e) {
		isMouseDown = false;
		console.log("d_mouseup", e);
	}
})();

/**
 * TODO APP
 */
(function () {
	var templateData = {
		"data": {
			"heading": "My Todo App",
			"newTodoItem": "",
			"todoItems": {},
			"isChecked": false
		},
		"functions": {
			"add_click": add_click,
			"newTodoItem_blur": newTodoItem_blur,
			"todo_dblclick": todo_dblclick,
			"checkbox_click": checkbox_click
		}
	};
	var appMountEl = document.getElementById("todoApp");
	var _template = "";
	var _context;

	function newTodoItem_blur(e) {
		templateData.data.newTodoItem = this.value;
		_context.update(templateData);
	}

	function add_click(e) {
		console.log("add item", templateData.data.newTodoItem);
		templateData.data.todoItems[guid()] = { "name": templateData.data.newTodoItem, "isDone": false };
		templateData.data.newTodoItem = "";
		_context.update(templateData);
	}

	function todo_dblclick(e) {
		templateData.data.todoItems[this.attributes["data-todo-id"].value].isDone = !templateData.data.todoItems[this.attributes["data-todo-id"].value].isDone;
		_context.update(templateData);
	}

	function checkbox_click(e) {
		templateData.data.isChecked = this.checked;
		_context.update(templateData);
	}

	function guid() {
		function S4() {
			return (((1+Math.random())*0x10000)|0).toString(16).substring(1); 
		}
		 
		return (S4() + S4() + "-" + S4() + "-4" + S4().substr(0,3) + "-" + S4() + "-" + S4() + S4() + S4()).toLowerCase();
	}

	function run() {
		fetch("todo.html").then(function(response) {
			return response.text();
		}).then(function (template) {
			_template = template;
			_context = __6oz.applyToDOM(appMountEl, _template, templateData);
		}).catch(function (err) {
			console.error("Couldn't fetch todo template", err);
		});
	}

	run();
})();

/**
 * COMPONENT APP
 */
(function () {
	var templateData = {
		"data": {
			"heading": "Component Test",
			"com1": [
				{ "heading": "Hello world", "text": "this is the first com1" },
				{ "heading": "Damn Daniel", "text": "Back at it again!" },
				{ "heading": "In the", "text": "black and white" },
				{ "heading": "Vans!", "text": "Haha" }
			],
			"com1Heading_click": function () {
				alert("com1 heading has been clicked");
			},
			"com2_click": function () {
				alert("com2 has been clicked");
			}
		},
		"functions": {
			"heading_click": heading_click,
			"com1Heading_click": function () {
				alert("com1 heading has been clicked");
			},
			"com2_click": function () {
				alert("com2 has been clicked");
			}
		}
	};
	var appMountEl = document.getElementById("componentApp");
	var _template = "";
	var _context;
	var _components = {
		"com1": {
			"componentName": "custom-component-1",
			"template": '<h6 onclick="props.com1Heading_click">Custom Component 1 - <%=props.heading %><span onclick="controller.custom_click">(<%=controller.text %>)</span></h6><p><%=props.text %></p><custom-component-2 com2_click={props.com2_click} text="<%=props.text %> - com2 subby" />',
			"controller": function () {
				this.text = "Hello from the controller";
				this.custom_click = function () {
					alert("This is a custom event from the controller");
				};
			}
		},
		"com2": {
			"componentName": "custom-component-2",
			"template": '<div onclick="props.com2_click"><pre>This is the sub component (com 2)</pre></div><p><%=props.text %></p>'
		}
	};

	function heading_click(e) {
		var com1 = templateData.data.com1;
		var reverseText = function (text) {
			return text.split("").reverse().join("");
		}

		for (var i = 0, l = com1.length; i < l; i++) {
			var c = com1[i];

			c.heading = reverseText(c.heading);
			c.text = reverseText(c.text);
		}

		_context.update(templateData);
	}

	function guid() {
		function S4() {
			return (((1+Math.random())*0x10000)|0).toString(16).substring(1); 
		}
		 
		return (S4() + S4() + "-" + S4() + "-4" + S4().substr(0,3) + "-" + S4() + "-" + S4() + S4() + S4()).toLowerCase();
	}

	function run() {
		__6oz.addComponent(_components.com1);
		__6oz.addComponent(_components.com2);

		fetch("componentApp.html").then(function(response) {
			return response.text();
		}).then(function (template) {
			_template = template;
			_context = __6oz.applyToDOM(appMountEl, _template, templateData);
		}).catch(function (err) {
			console.error("Couldn't fetch todo template", err);
		});
	}

	run();
})();

src/6oz.js
/**
 * v1.0.2
 * 6oz because coffee
      ( (
      ) )
    ........
    |      |]
    \      /
     `----'
 */

/*! loglevel - v1.4.1 - https://github.com/pimterry/loglevel - (c) 2016 Tim Perry - licensed MIT */
!function(a,b){"use strict";"function"==typeof define&&define.amd?define(b):"object"==typeof module&&module.exports?module.exports=b():a.log=b()}(this,function(){"use strict";function a(a){return typeof console===h?!1:void 0!==console[a]?b(console,a):void 0!==console.log?b(console,"log"):g}function b(a,b){var c=a[b];if("function"==typeof c.bind)return c.bind(a);try{return Function.prototype.bind.call(c,a)}catch(d){return function(){return Function.prototype.apply.apply(c,[a,arguments])}}}function c(a,b,c){return function(){typeof console!==h&&(d.call(this,b,c),this[a].apply(this,arguments))}}function d(a,b){for(var c=0;c<i.length;c++){var d=i[c];this[d]=a>c?g:this.methodFactory(d,a,b)}}function e(b,d,e){return a(b)||c.apply(this,arguments)}function f(a,b,c){function f(a){var b=(i[a]||"silent").toUpperCase();try{return void(window.localStorage[l]=b)}catch(c){}try{window.document.cookie=encodeURIComponent(l)+"="+b+";"}catch(c){}}function g(){var a;try{a=window.localStorage[l]}catch(b){}if(typeof a===h)try{var c=window.document.cookie,d=c.indexOf(encodeURIComponent(l)+"=");d&&(a=/^([^;]+)/.exec(c.slice(d))[1])}catch(b){}return void 0===k.levels[a]&&(a=void 0),a}var j,k=this,l="loglevel";a&&(l+=":"+a),k.levels={TRACE:0,DEBUG:1,INFO:2,WARN:3,ERROR:4,SILENT:5},k.methodFactory=c||e,k.getLevel=function(){return j},k.setLevel=function(b,c){if("string"==typeof b&&void 0!==k.levels[b.toUpperCase()]&&(b=k.levels[b.toUpperCase()]),!("number"==typeof b&&b>=0&&b<=k.levels.SILENT))throw"log.setLevel() called with invalid level: "+b;return j=b,c!==!1&&f(b),d.call(k,b,a),typeof console===h&&b<k.levels.SILENT?"No console available for logging":void 0},k.setDefaultLevel=function(a){g()||k.setLevel(a,!1)},k.enableAll=function(a){k.setLevel(k.levels.TRACE,a)},k.disableAll=function(a){k.setLevel(k.levels.SILENT,a)};var m=g();null==m&&(m=null==b?"WARN":b),k.setLevel(m,!1)}var g=function(){},h="undefined",i=["trace","debug","info","warn","error"],j=new f,k={};j.getLogger=function(a){if("string"!=typeof a||""===a)throw new TypeError("You must supply a name when creating a logger.");var b=k[a];return b||(b=k[a]=new f(a,j.getLevel(),j.methodFactory)),b};var l=typeof window!==h?window.log:void 0;return j.noConflict=function(){return typeof window!==h&&window.log===j&&(window.log=l),j},j});

var __escapeHTML = function (rawHTML) {
	if (!rawHTML) {
		return "";
	}

	if (typeof rawHTML != "string") {
		rawHTML = rawHTML.toString();
	}

	return rawHTML
		.replace(/&/g, "&amp;")
		.replace(/</g, "&lt;")
		.replace(/>/g, "&gt;")
		.replace(/"/g, "&quot;")
		.replace(/'/g, "&apos;");
};
var __unescapeHTML = function (rawValue) {
	if (!rawValue) {
		return "";
	}

	if (typeof rawValue != "string") {
		rawValue = rawValue.toString();
	}

	return rawValue
		.replace(/&amp;/g, "&")
		.replace(/&lt;/g, "<")
		.replace(/&gt;/g, ">")
		.replace(/&quot;/g, "\"")
		.replace(/&apos;/g, "'");
};

(function () {
	"use strict";
	/**
	 * EXTERNAL LIBS/HELPERS
	 */
	//COURTESY OF moonjs.ga
	var __lex = (function () {
		// Concatenation Symbol
		var concatenationSymbol = " + ";
		// Opening delimiter
		var openRE = /\{\{\s*/;
		// Closing delimiter
		var closeRE = /\s*\}\}/;
		// Whitespace character
		var whitespaceCharRE = /[\s\n]/;
		// All whitespace
		var whitespaceRE = /[\s\n]/g;
		// Start of a tag or comment
		var tagOrCommentStartRE = /<\/?(?:[A-Za-z]+\w*)|<!--/;
		// Dynamic expressions
		var expressionRE = /"[^"]*"|'[^']*'|\d+[a-zA-Z$_]\w*|\.[a-zA-Z$_]\w*|[a-zA-Z$_]\w*:|([a-zA-Z$_]\w*)(?:\s*\()?/g;
		// HTML Escapes
		var escapeRE = /(?:(?:&(?:amp|gt|lt|nbsp|quot|apos|#39);)|'|"|\\|\n)/g;
		var escapeMap = {
			"&amp;": '&',
			"&gt;": '>',
			"&lt;": '<',
			"&nbsp;": ' ',
			"&quot;": "\\\"",
			"&apos;": "\\\'",
			"&#39;": "\\\'",
			"'": "\\\'",
			'\\': "\\\\",
			'"': "\\\"",
			'\n': "\\n"
		};
		// Global Variables/Keywords
		var globals = ["instance", "staticNodes", "true", "false", "undefined", "null", "NaN", "typeof", "in"];
		// Void and SVG Elements
		var VOID_ELEMENTS = ["area", "base", "br", "command", "embed", "hr", "img", "input", "keygen", "link", "meta", "param", "source", "track", "wbr"];
		var SVG_ELEMENTS = ["animate", "circle", "clippath", "cursor", "defs", "desc", "ellipse", "filter", "font-face", "foreignObject", "g", "glyph", "image", "line", "marker", "mask", "missing-glyph", "path", "pattern", "polygon", "polyline", "rect", "svg", "switch", "symbol", "text", "textpath", "tspan", "use", "view"];
		// Data Flags
		var FLAG_SVG = 1;
		var FLAG_STATIC = 1 << 1;
		// Trim Whitespace
		var trimWhitespace = function (value) {
			return value.replace(whitespaceRE, '');
		};

		return function (template) {
			var length = template.length;
			var tokens = [];
			var current = 0;
		
			while (current < length) {
				var char = template[current];

				if (char === '<') {
					current++;

					if (template.substring(current, current + 3) === "!--") {
						// Comment
						current += 3;
						var endOfComment = template.indexOf("-->", current);

						if (endOfComment === -1) {
							current = length;
						} else {
							current = endOfComment + 3;
						}
					} else {
						// Tag
						var tagToken = {
							type: "Tag",
							value: ''
						};
				
						var tagType = '';
						var attributes = {};
						var closeStart = false;
						var closeEnd = false;
				
						char = template[current];
				
						// Exit starting closing slash
						if (char === '/') {
							char = template[++current];
							closeStart = true;
						}
				
						// Get tag name
						while ((current < length) && ((char !== '>') && (char !== '/') && (whitespaceCharRE.test(char) === false))) {
							tagType += char;
							char = template[++current];
						}
				
						// Iterate to end of tag
						while ((current < length) && ((char !== '>') && (char !== '/' || template[current + 1] !== '>'))) {
							if (whitespaceCharRE.test(char) === true) {
								// Skip whitespace
								char = template[++current];
							} else {
							// Find attribute name
								var attrName = '';
								var attrValue = '';

								while ((current < length) && ((char !== '=') && (whitespaceCharRE.test(char) === false) && ((char !== '>') && (char !== '/' || template[current + 1] !== '>')))) {
									attrName += char;
									char = template[++current];
								}

								// Find attribute value
								if (char === '=') {
									char = template[++current];
					
									var quoteType = ' ';
									var isScopedValue = false;
									var openBraceCount = 0;

									if (char === '"' || char === '\'' || char === ' ' || char === '\n' || char === '{') {
										isScopedValue = char === '{';
										quoteType = isScopedValue ? '}' : char;
										char = template[++current];
									}

									if (isScopedValue) {
										attrValue += '{';
									}

									// Iterate to end of quote type, or end of tag
									while ((current < length) && ((char !== '>') && (char !== '/' || template[current + 1] !== '>'))) {
										if ((!isScopedValue || (isScopedValue && openBraceCount <= 0)) && char === quoteType) {
											if (isScopedValue) {
												attrValue += '}';
											}

											char = template[++current];
											break;
										} else {
											attrValue += char;
											
											if (char === '{') {
												openBraceCount++;
											} else if (char === '}') {
												openBraceCount--;
											}

											char = template[++current];
										}
									}
								}
					
								var attrToken = {
									name: attrName,
									value: __unescapeHTML(attrValue),
									argument: undefined,
									data: {}
								};
					
								var splitAttrName = attrName.split(':');

								if(splitAttrName.length === 2) {
									attrToken.name = splitAttrName[0];
									attrToken.argument = splitAttrName[1];
								}
					
								attributes[attrName] = attrToken;
							}
						}
				
						if (char === '/') {
							current += 2;
							closeEnd = true;
						} else {
							current++;
						}
				
						tagToken.value = tagType;
						tagToken.attributes = attributes;
						tagToken.closeStart = closeStart;
						tagToken.closeEnd = closeEnd;
						tokens.push(tagToken);
					}
				} else {
					// Text
					var textTail = template.substring(current);
					var endOfText = textTail.search(tagOrCommentStartRE);
					var text;

					if (endOfText === -1) {
						text = textTail;
						current = length;
					} else {
						text = textTail.substring(0, endOfText);
						current += endOfText;
					}

					if (trimWhitespace(text).length !== 0) {
						tokens.push({
							type: "Text",
							value: text.replace(escapeRE, function (match) {
								return escapeMap[match];
							})
						});
					}
				}
			}

			return tokens;
		} 
	})();
	//COURTESY OF MR. J. RESIG
	var __tmpl = (function () {
		var cache = {};
		var DELIMITER = { "start": "<%", "end": "%>" };

		//private functions
		return function tmpl(str, data) {  // Simple JavaScript Templating || John Resig - http://ejohn.org/ - MIT Licensed
			// Figure out if we're getting a template, or if we need to
			// load the template - and be sure to cache the result.
			var fn = !/\W/.test(str) ?
				cache[str] = cache[str] ||
				tmpl(document.getElementById(str).innerHTML) :

				// Generate a reusable function that will serve as a template
				// generator (and which will be cached).
				new Function("obj",
					"var p=[],print=function(){p.push.apply(p,arguments);};" +
					//PM - Added fix for issue where null obj throws an error
					"obj = obj || {};" +

					// Introduce the data as local variables using with(){}
					"with(obj){p.push('" +

					// Convert the template into pure JavaScript
					// Grabbed apostrophe fix from http://weblog.west-wind.com/posts/2008/Oct/13/Client-Templating-with-jQuery
					str
						.replace(/[\r\t\n]/g, " ")
						.replace(new RegExp("'(?=[^" + DELIMITER.end.substr(0, 1) + "]*" + DELIMITER.end + ")", "g"), "\t")
						.split("'").join("\\'")
						.replace(new RegExp("=\{(.+?(\}+)?)\}", "g"), "=',__6oz.registerData($1),'")
						.split("\t").join("'")
						.replace(new RegExp(DELIMITER.start + ":(.+?)" + DELIMITER.end, "g"), "',__escapeHTML($1),'")
						.replace(new RegExp(DELIMITER.start + "=(.+?)" + DELIMITER.end, "g"), "',$1,'")
						.split(DELIMITER.start).join("');")
						.split(DELIMITER.end).join("p.push('")
					+ "');}return p.join('');"
				);

			// Provide some basic currying to the user
			return data ? fn(data) : fn;
		};
	})();

	/**
	 * LIBRARY STARTS HERE
	 */
	var isFunctionRE = /^on[a-z]+$/i;
	var isPropRE = /^props$/i;
	var PARAM_NAME = "__data";
	var LOG_PREFIX = ">>[6oz.js] ";
	var _lib = {};
	var _logger = log.getLogger("6oz logger");
	var _components = {};
	var _dataRegister = {};
	var _componentProps = {};
	var _dataToIncrementalDOM = {};

	function applyToDOM(el, template, templateData) {
		var renderedTemplate = __tmpl(template, templateData.data);
		var lexResults = __lex(renderedTemplate);
		var renderFunctionBody = processLexTemplate(lexResults, templateData.functions, PARAM_NAME + ".__fn");
		var renderFunction = new Function(PARAM_NAME, renderFunctionBody.join(""));

		_dataToIncrementalDOM.__fn = templateData.functions;
		IncrementalDOM.patch(el, renderFunction, _dataToIncrementalDOM);
		_dataRegister = {};
		_dataToIncrementalDOM = {};
		return {
			"update": function (updateTemplateData) {
				applyToDOM(el, template, updateTemplateData);
			}
		};
	}
	function processLexTemplate(lexResults, templateFunctions, functionPrefix) {
		var renderFunctionBody = [];
		var isComponentTag = function (tagName) {
			return tagName && tagName.indexOf("-") > -1;
		};
		var nodeLevel = 0;
		var weAreSkippingNodes = false;

		for (var i = 0, l = lexResults.length; i < l; i++) {
			var cLex = lexResults[i];
			var cLexType = cLex.type;
			var cLexValue = cLex.value;
			var thisIsComponentTag = isComponentTag(cLexValue);
			var isSkipNode = cLex.attributes && cLex.attributes["skip-node"] && cLex.attributes["skip-node"].value == "true";
			var getTagDetails = function (c, t, f) {
				var results = processTag(c, t, f);

				results.elementKeyID = results.elementKeyID == null ? "null" : '"' + results.elementKeyID + '"';
				return results;
			};
			var tagDetails;

			if (cLexType == "ComponentStart") {
				_dataToIncrementalDOM[cLex.attributes.componentID] = templateFunctions;
				renderFunctionBody.push('(function (o) {');
			} else if (cLexType == "ComponentEnd") {
				renderFunctionBody.push('})(' + PARAM_NAME + '["' + cLex.attributes.componentID + '"]);');
			} else if (cLexType == "Tag") {
				if (!cLex.closeStart && isSkipNode && !weAreSkippingNodes) {
					tagDetails = getTagDetails(cLex, templateFunctions, functionPrefix);
					renderFunctionBody.push('IncrementalDOM.' + tagDetails.elementType + '("' + cLexValue + '",' + tagDetails.elementKeyID + ',null' + tagDetails.tagAttributesString + ');');
					renderFunctionBody.push("IncrementalDOM.skip();");
					weAreSkippingNodes = true;
					nodeLevel = 0;
				} else if (cLex.closeStart && !thisIsComponentTag) {
					if (nodeLevel-- == 0) {
						weAreSkippingNodes = false;
					}

					if (!weAreSkippingNodes) {
						renderFunctionBody.push('IncrementalDOM.elementClose("' + cLexValue + '");');
					}
				} else if (thisIsComponentTag && !weAreSkippingNodes) {
					var componentResults = processComponent(cLex);

					if (componentResults === false) {
						continue;
					}

					renderFunctionBody = renderFunctionBody.concat(componentResults);
				} else if (!thisIsComponentTag) {
					nodeLevel++;

					if (!weAreSkippingNodes) {
						tagDetails = getTagDetails(cLex, templateFunctions, functionPrefix);
						renderFunctionBody.push('IncrementalDOM.' + tagDetails.elementType + '("' + cLexValue + '",' + tagDetails.elementKeyID + ',null' + tagDetails.tagAttributesString + ');');
					}
				}
			} else if (cLexType == "Text" && !weAreSkippingNodes) {
				renderFunctionBody.push('IncrementalDOM.text("' + cLexValue.replace(/(?:\r\n|\r|\n)/g, "") + '");');
			}
		}

		return renderFunctionBody;
	}
	function processTag(cLex, templateFunctions, functionPrefix) {
		if (!cLex) {
			return false;
		}

		var attributes = cLex.attributes;
		var tagAttributes = [];
		var tagAttributesString = "";
		var elementType = cLex.closeEnd ? "elementVoid" : "elementOpen";
		var elementKeyID = null;
		var checkForFunction = function (templateFunctions, fnName) {
			if (fnName.indexOf(".") == -1) {
				return templateFunctions[fnName];
			}

			var fnNameSplit = fnName.split(".");
			var fnFound = templateFunctions;
			var madeItToEnd = true;

			for (var i = 0, l = fnNameSplit.length; i < l; i++) {
				var currentFn = fnFound[fnNameSplit[i]];

				if (!currentFn) {
					madeItToEnd = false;
					break;
				}

				fnFound = currentFn;
			}

			if (madeItToEnd) {
				return fnFound;
			}
		};

		if (Object.keys(attributes).length > 0) {
			for (var attributeKey in attributes) {
				var attribute = attributes[attributeKey];
				var attributeName = attribute.name;
				var attributeValue = attribute.value.replace(/(?:\r\n|\r|\n)/g, "");
				var isEventAttribute = isFunctionRE.test(attributeName);
				var isPropAttribute = isPropRE.test(attributeName);

				if (attributeName.toLowerCase() == "key-id") {
					elementKeyID = attributeValue;
					continue;
				}

				tagAttributes.push('"' + attribute.name + '"');

				if (isEventAttribute) {
					var candidateFunction = checkForFunction(templateFunctions, attributeValue);

					if (typeof(candidateFunction) === "function") {
						tagAttributes.push(functionPrefix + "." + attributeValue);
					} else {
						_logger.warn(LOG_PREFIX + "Could not find the function " + attributeValue + " so event was not bound.");
						tagAttributes.pop();
					}
				} else if (isPropAttribute) {
					if (!_dataToIncrementalDOM.attributeProps) {
						_dataToIncrementalDOM.attributeProps = {};
					}
					
					var attributeFinalValue = '"' + attribute.value + '"';
					var attributeGUID = /\{([a-zA-Z0-9\-]+)\}/.exec(attributeValue);

					if (attributeGUID && attributeGUID.length > 1) {
						attributeGUID = attributeGUID[1];
						_dataToIncrementalDOM.attributeProps[attributeGUID] = _dataRegister[attributeGUID];
						attributeFinalValue = PARAM_NAME + ".attributeProps['" + attributeGUID + "']";
					}

					tagAttributes.push(attributeFinalValue);
				} else {
					tagAttributes.push('"' + attributeValue.replace(/"/g, "\\\"") + '"');
				}
			}
		}

		tagAttributesString = tagAttributes.length > 0 ? ',' + tagAttributes.join(',') : "";
		return { "elementType": elementType, "tagAttributesString": tagAttributesString, "elementKeyID": elementKeyID };
	}
	function processComponent(cLex) {
		var componentName = cLex.value.toLowerCase();
		var componentDetails = _components[componentName];
		
		if (!componentDetails) {
			return false;
		}

		var cLexAttributes = cLex.attributes;
		var attributes = {};
		var checkForScopedValue = function (value) {
			return value && /^\{(.*)\}$/.exec(value);
		};

		for (var key in cLexAttributes) {
			var attributeValue = cLexAttributes[key].value;
			var scopedValueResult = checkForScopedValue(attributeValue);

			if (scopedValueResult && scopedValueResult.length > 1 && scopedValueResult[1]) {
				var scopedValue = scopedValueResult[1];
				
				if (_dataRegister[scopedValue]) {
					attributes[key] = _dataRegister[scopedValue];
				} else {
					attributes[key] = undefined;
				}
			} else {
				attributes[key] = attributeValue;
			}
		}

		var componentController;
		var templateData = {
			"props": attributes
		};

		if (componentDetails.controller !== undefined) {
			componentController = new componentDetails.controller(attributes);
			templateData.controller = componentController;
		}

		var renderedTemplate = __tmpl(componentDetails.template, templateData);
		var lexResults = __lex(renderedTemplate);
		var componentID = guid();

		lexResults.unshift({ "type": "ComponentStart", "attributes": { "componentID": componentID } });
		lexResults.push({ "type": "ComponentEnd", "attributes": { "componentID": componentID } });
		attributes.props = {};

		for (var attributeName in attributes) {
			attributes.props[attributeName] = attributes[attributeName];
		}

		if (componentController !== undefined) {
			attributes.controller = componentController;
		}

		return processLexTemplate(lexResults, attributes, "o");
	}
	function addComponent(o) {
		o = o || {};

		if (!o.componentName || !o.template) {
			_logger.warn(LOG_PREFIX + "You must provide a componentName and template.");
			return false;
		}

		var componentName = o.componentName.toLowerCase();
		var controller = o.controller;
		var controllerToLoad;

		if (!/-[a-zA-Z0-9]+[^-]?$/.test(componentName)) {
			_logger.warn(LOG_PREFIX + "You must provide a valid component name [" + componentName + "].  It must be kebab case, must contain at least one hyphen, and must not end with a hyphen.");
			return;
		}

		if (_components[componentName]) {
			_logger.warn(LOG_PREFIX + "A component with name " + componentName + " has already been added.");
			return;
		}

		if (controller !== undefined && typeof(controller) === "function") {
			controllerToLoad = controller;
		}

		_components[componentName] = {
			"template": o.template,
			"controller": controllerToLoad
		};
		return true;
	}
	function registerData(data) {
		var dataKey = guid();

		_dataRegister[dataKey] = data;
		return "{" + dataKey + "}";
	}
	function guid() {
		function S4() {
			return (((1+Math.random())*0x10000)|0).toString(16).substring(1); 
		}

		return (S4() + S4() + "-" + S4() + "-4" + S4().substr(0,3) + "-" + S4() + "-" + S4() + S4() + S4()).toLowerCase();
	}

	_lib.applyToDOM = applyToDOM;
	_lib.addComponent = addComponent;
	_lib.registerData = registerData;

	window.__6oz = _lib;

	IncrementalDOM.attributes.props = function (el, name, value) {
		el[name] = value;
	};
})();
